### PR TITLE
feat(core): handle flags in autocomplete

### PR DESCRIPTION
**Pull request by loicbourgois**

### CODE REVIEW

File: internal/core/autocomplete.go [hunk 1]
```diff
@@ -116,11 +173,27 @@ func AutoComplete(ctx context.Context, leftWords []string, wordToComplete string
...
+
+		case !exist:
+			// we did not reach a leaf command, and word is unknown
+			return &AutocompleteResponse{}
+
+		case children.Type == AutoCompleteNodeTypeArgument:
+			// Do nothing
+
+		default:
+			// word is a namespace or verb or resource or flag or flag value
```
Comment (QuentinBrosse) [inline internal/core/autocomplete.go]: nit: `Word`

File: internal/core/autocomplete.go [hunk 2]
```diff
@@ -116,11 +173,27 @@ func AutoComplete(ctx context.Context, leftWords []string, wordToComplete string
...
+		switch {
+		case !exist && node.isLeafCommand():
+			// word is probably an unknown argument
+			// Just skip it
+
+		case !exist:
+			// we did not reach a leaf command, and word is unknown
+			return &AutocompleteResponse{}
+
+		case children.Type == AutoCompleteNodeTypeArgument:
```
Comment (QuentinBrosse) [inline internal/core/autocomplete.go]: Add a comment explaining why you do nothing for an argument (but continue to dig for flag)

File: internal/core/autocomplete.go [hunk 3]
```diff
@@ -28,8 +37,20 @@ type AutoCompleteNode struct {
 	Children map[string]*AutoCompleteNode
 	Command  *Command
 	ArgSpec  *ArgSpec
-	IsArg    bool
-	// TODO: support flags in the autocomplete node
+	Type     AutoCompleteNodeType
+}
+
+func (n *AutoCompleteNode) addGlobalFlags() {
```
[the remark this answers is not in the document]
Comment (loicbourgois) [reply internal/core/autocomplete.go]: created ticket APIGW-1758

File: internal/core/autocomplete.go [hunk 4]
```diff
@@ -53,7 +75,21 @@ func NewAutoCompleteNode() *AutoCompleteNode {
 func NewArgAutoCompleteNode(argSpec *ArgSpec) *AutoCompleteNode {
 	node := NewAutoCompleteNode()
 	node.ArgSpec = argSpec
-	node.IsArg = true
+	node.Type = AutoCompleteNodeTypeArgument
+	return node
+}
+
+func NewFlagAutoCompleteNode(parent *AutoCompleteNode, values []string) *AutoCompleteNode {
```
[the remark this answers is not in the document]
Comment (loicbourgois) [reply internal/core/autocomplete.go]: >Does values == possibleValues (of the flag)?
yes

>We happen for a flag who can take any values (eg. --secret-key)?
actually, it is not supported yet
we should add it in this PR

File: internal/core/autocomplete.go [hunk 5]
```diff
@@ -116,11 +173,27 @@ func AutoComplete(ctx context.Context, leftWords []string, wordToComplete string
...
 	node := commandTreeRoot
-	for i, word := range leftWords {
+	for i := 0; i < len(leftWords); i++ {
+		word := leftWords[i]
 		children, exist := node.Children[word]
-		if i < len(leftWords) && exist && !children.IsArg {
+
+		switch {
+		case !exist && node.isLeafCommand():
+			// word is probably an unknown argument
```
[the remark this answers is not in the document]
Comment (loicbourgois) [reply internal/core/autocomplete.go]: won't do: `word` is the variable name

File: internal/core/autocomplete.go [hunk 6]
```diff
@@ -53,7 +75,21 @@ func NewAutoCompleteNode() *AutoCompleteNode {
 func NewArgAutoCompleteNode(argSpec *ArgSpec) *AutoCompleteNode {
 	node := NewAutoCompleteNode()
 	node.ArgSpec = argSpec
-	node.IsArg = true
+	node.Type = AutoCompleteNodeTypeArgument
+	return node
```
Comment (kindermoumoute) [inline internal/core/autocomplete.go]: nit: 
```
return &AutoCompleteNode{
		Children: map[string]*AutoCompleteNode{},
                ArgSpec: argSpec
		Type:     AutoCompleteNodeTypeArgument,
	}
```

File: internal/core/autocomplete.go [hunk 7]
```diff
@@ -45,6 +66,7 @@ func newAutoCompleteResponse(suggestions []string) *AutocompleteResponse {
 func NewAutoCompleteNode() *AutoCompleteNode {
```
Comment (kindermoumoute) [inline internal/core/autocomplete.go]: nit: NewAutoCompleteCommandNode

File: internal/core/autocomplete.go [hunk 8]
```diff
@@ -53,7 +75,21 @@ func NewAutoCompleteNode() *AutoCompleteNode {
 func NewArgAutoCompleteNode(argSpec *ArgSpec) *AutoCompleteNode {
```
Comment (kindermoumoute) [inline internal/core/autocomplete.go]: NewAutoCompleteArgNode

File: internal/core/autocomplete.go (hunk 4 above)
Comment (kindermoumoute) [inline internal/core/autocomplete.go]: NewAutoCompleteFlagNode

File: internal/core/autocomplete.go [hunk 9]
```diff
@@ -53,7 +75,21 @@ func NewAutoCompleteNode() *AutoCompleteNode {
...
 	node := NewAutoCompleteNode()
 	node.ArgSpec = argSpec
-	node.IsArg = true
+	node.Type = AutoCompleteNodeTypeArgument
+	return node
+}
+
+func NewFlagAutoCompleteNode(parent *AutoCompleteNode, values []string) *AutoCompleteNode {
+	node := NewAutoCompleteNode()
```
Comment (kindermoumoute) [inline internal/core/autocomplete.go]: inline struct
```
node := &AutoCompleteNode{
		Children: map[string]*AutoCompleteNode{},
		Type:     AutoCompleteNodeTypeFlag,
	}
```

File: internal/core/autocomplete.go (hunk 4 above)
Comment (kindermoumoute) [inline internal/core/autocomplete.go]: input parameters should be (parent *AutoCompleteNode, values ...string)

File: internal/core/autocomplete.go [hunk 10]
```diff
@@ -53,7 +75,21 @@ func NewAutoCompleteNode() *AutoCompleteNode {
...
+	node.Type = AutoCompleteNodeTypeFlag
+	for _, v := range values {
+		node.Children[v] = NewAutoCompleteNode()
```
Comment (kindermoumoute) [inline internal/core/autocomplete.go]: inline struct

File: internal/core/autocomplete.go [hunk 11]
```diff
@@ -28,8 +37,20 @@ type AutoCompleteNode struct {
 	Children map[string]*AutoCompleteNode
 	Command  *Command
 	ArgSpec  *ArgSpec
-	IsArg    bool
-	// TODO: support flags in the autocomplete node
+	Type     AutoCompleteNodeType
+}
+
+func (n *AutoCompleteNode) addGlobalFlags() {
+	n.Children["--access-key"] = NewFlagAutoCompleteNode(n, nil)
```
Comment (kindermoumoute) [inline internal/core/autocomplete.go]: can we add this at the same level as flag declaration?

File: internal/core/autocomplete_test.go [hunk 12]
```diff
@@ -118,6 +121,15 @@ func TestAutocomplete(t *testing.T) {
 	t.Run("scw test flower create leaves.0.size=M leaves", run(&testCase{Suggestions: AutocompleteSuggestions{"leaves.1.size="}}))
 	t.Run("scw test flower create leaves.0.size=M leaves leaves.1.size=M", run(&testCase{WordToCompleteIndex: 5, Suggestions: AutocompleteSuggestions{"leaves.2.size="}}))
 	// TODO: t.Run("scw test flower create leaves.0.size=", run(&testCase{Suggestions: AutocompleteSuggestions{"L", "M", "S", "XL", "XXL"}}))
+
+	t.Run("scw -", run(&testCase{Suggestions: AutocompleteSuggestions{"--access-key", "--debug", "--help", "--output", "--profile", "--secret-key", "-D", "-h", "-o", "-p"}}))
+	t.Run("scw test -o j", run(&testCase{Suggestions: AutocompleteSuggestions{"json"}}))
+	t.Run("scw test flower -o ", run(&testCase{Suggestions: AutocompleteSuggestions{"human", "json"}}))
+	t.Run("scw test flower -o json create -", run(&testCase{Suggestions: AutocompleteSuggestions{"--access-key", "--debug", "--help", "--output", "--profile", "--secret-key", "--wait", "-D", "-h", "-p", "-w"}}))
```
Comment (kindermoumoute) [inline internal/core/autocomplete_test.go]: --output should not appear

Comment (loicbourgois) [reply internal/core/autocomplete_test.go]: Internal ticket: APIGW-1757

File: internal/core/autocomplete.go (hunk 11 above)
Comment (loicbourgois) [reply internal/core/autocomplete.go]: Internal ticket: APIGW-1758

File: internal/core/autocomplete.go [hunk 13]
```diff
@@ -72,20 +78,36 @@ func NewAutoCompleteNode() *AutoCompleteNode {
...
-	for _, v := range values {
-		node.Children[v] = NewAutoCompleteNode()
-		node.Children[v].Children = parent.Children
-		node.Children[v].Type = AutoCompleteNodeTypeFlagValue
+// NewFlagAutoCompleteNode returns a node representing a Flag.
+// It creates the children node with possible values if they exist.
+// It sets parent.children as children of the lowest nodes:
+//  the lowest node is the flag if it has no possible value ;
+//  or the lowest nodes are the possible values if the exist.
+func NewAutoCompleteFlagNode(name string, parent *AutoCompleteNode, hasVariableValue bool, possibleValues ...string) *AutoCompleteNode {
```
Comment (kindermoumoute) [inline internal/core/autocomplete.go]: Should take a `FlagSpec` struct

```
NewAutoCompleteFlagNode(&FlagSpec{
    Name string
    IsBool bool
    EnumValues []string
})
```

File: internal/core/autocomplete.go [hunk 14]
```diff
@@ -72,20 +78,36 @@ func NewAutoCompleteNode() *AutoCompleteNode {
...
+	node := &AutoCompleteNode{
+		Children: map[string]*AutoCompleteNode{},
```
Comment (kindermoumoute) [inline internal/core/autocomplete.go]: nit: use `make(map[string]*AutoCompleteNode)`

**Pull request merged**